Fix observation of deposits to be settled

ch1bo opened the pull request:
This is will be the last PR related to deposits that also fixes the "deposit used too early" problem identified in course of #1951 and related discussions. With all the prior work already done in #1969, #1974 and the refactorings of #1977 this was fairly straight forward. 

:alarm_clock: Breaking change on the network as I needed to change the `ReqSn` message.

:alarm_clock: Needed to substantially increase the `defaultTTL` of re-enqueing inputs. This is needed because the `ReqSn` handling will wait for the deposit to become active (passes the `--deposit-period`) and needs to act on the `ReqSn` only then. This is problematic because the input queue is not (yet?) persisted and restarting the node will lose that state -> the deposit will become active, but an `AckSn` would not be sent. See https://github.com/cardano-scaling/hydra/issues/1999 for a follow-up to address this.

:alarm_clock: The fact that we observe a `SlotNo` in `DepositObservation` requires a `TimeHandle` to convert it further to a `UTCTime` (in `convertObservation`). As the `hydra-chain-observer` does not have a `TimeHandle`, I decided to switch to using `HeadObservation` (the type from `hydra-tx`) instead of `OnChainTx` (the type from `hydra-node`) for the chain observer / `hydra-explorer` interface. This results in quite a lot of (backwards compatible) changes and this companion PR: https://github.com/cardano-scaling/hydra-explorer/pull/47

TODO:
- [x] spec changes https://github.com/cardano-scaling/hydra-formal-specification/pull/18
- [x] make chain observers not break the explorer API

---

* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter
  - Two new TODOs in `Hydra.HeadLogic.onOpenNetworkReqSn` on things we're not sure that are needed or even missing from the implementation (not critical, but just inconsistent)

ch1bo committed:
Add a test for deposit settlement

ch1bo committed:
Add an upper validity to depositTx

This will mark the creation of the deposit.

ch1bo committed:
Convert deposit creation slot into time

This changes the convertObservation function to require a TimeHandle and
will make chain observers need to use a different type than OnChainTx.

ch1bo committed:
Fix deposit approval before deposit period

This handles the created time on Deposit to determine whether it is
still inactive or can be requested / signed upon a snapshot request.

Importantly, the snapshot signing must be able to wait for the deposit
to become active and then submit snapshot acknowledgement. To that end,
the TTL and time between re-enqueuing waits is greatly increased now.

ch1bo committed:
Add DepositActivated server output

This was helpful during debugging (until we have a proper GET /deposits).

ch1bo committed:
Switch ChainObservation to use HeadObservation

This will allow chain observers to use these valueu directly without
further conversion to 'OnChainTx'. The latter will be the hydra-node
specific type which contains already converted time fields.

The change here is incomplete as HeadObservation is missing several
instances and breaks the explorer API.

v0d1ch reviewed (commented):
It feels like we should make sure we all understand the impact of these recent changes especially if you will not be around so much in the following days. I find it personally hard to review large PRs (previous stacked PR) without having a dedicated session with the author (which is fine since you plan to do that today)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -102,7 +102,7 @@ import Hydra.Tx.Party (Party (vkey))
 
 defaultTTL :: TTL
-defaultTTL = 5
+defaultTTL = 10000
```
Must: only have one defaultTTL

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -310,9 +310,13 @@ stepHydraNode node = do
+-- | The time to wait between re-enqueuing a 'Wait' outcome
 waitDelay :: DiffTime
-waitDelay = 0.1
+waitDelay = 1
```
This is not good, we should _at least_ make this wait configurable different for waits "on L2 stuff" vs. the wait for deposits to become active.

However, we should address the situation of protocol logic needing to wait in the first place!

ch1bo commented:
In course of a discussion session we identified that even on `master` we have an issue with too low `wait` times in at least this situation:

![image](https://github.com/user-attachments/assets/1ab9ede6-69ed-4e3e-99b5-24b8dc1e870a)

When a party that is not the snapshot leader requests a decrement via `ReqDec`, there must not be a commit/increment $U_\alpha$ pending. Even without the changes of this PR, increments may take several seconds (`~20s` blocks on mainnet) while the `defaultTTL` and delay between retries upon a `wait` outcome results in a maximum wait time of `0.5s`.

Hence this PR does not worsen the situation, but we should make the `Wait` outcome parameterizable so we can indicate how to long keep an event in the queue. Also, a follow-up issue / PR could investigate whether switching the `waitDeposit` in `ReqSn` handling deliberately to a require and retry snapshot requests on each `Tick` would be a path forward. 

In general, the protocol logic feels more and more contrived and sending requests that may need to be re-enqueued is not very failure resistant / good practice in distributed systems.

ch1bo committed:
Accept longer re-enqueueing and add known issue

ch1bo committed:
Simplify InitObservation

ch1bo committed:
Simplify CommitObservation

The note about observing without using resolved inputs was outdated and
the approach now was deemed fine in the past (contrary to the wording
used). See commit 91d48d869a85d3e7b9a2c92ab72d4fa242545d43

ch1bo committed:
Simplify CollectComObservation

The Direct.State module is lots of dead code. The OpenState was not
actually using the OpenThreadOutput other than for getKnownUTxO.

ch1bo committed:
Simplify CloseObservation

ch1bo committed:
Move {Open,Closed}ThreadOutput types

They are now only used as parameters to close and contest transaction
creation respectively.

ch1bo committed:
Simplify ContestObservation

Turns out the contestedThreadOutput was not used at all

ch1bo committed:
Draft JSON instances for HeadObservation

ch1bo committed:
Move genMintedOrBurnedValue

ch1bo committed:
Change OnDepositTx to not use POSIXTime

ch1bo committed:
Move Arbitrary instances of HeadObservation types

ch1bo committed:
Generate HeadObservation golden files

ch1bo committed:
Align InitObservatin further with OnInitTx

This allows us to define a backwards compatible JSON parser only using
the HeadObservation type.

ch1bo committed:
Custom HeadObservation serialization and switch golden files

We want a serialization that resembles OnChainTx so the hydra-explorer
is backwards compatible (tested there).

ch1bo committed:
Bump versions and update changelog

ch1bo committed:
Remove unused instances

Thanks weeder

v0d1ch commented on docs/docs/known-issues.md:
```diff
@@ -57,6 +57,10 @@ Note that, as with any adjustments of this kind, it is good practice to make a b
 
+### Deposit periods
+
+The `--deposit-period` allows an individual `hydra-node` operator to decide how long they want a deposit to have settled at least. However, differences bigger than [`defaultTTL * waitDelay`](https://hydra.family/head-protocol/haddock/hydra-node/Hydra-Node.html#v:waitDelay) (currently 10 minutes) result in non-approved snapshots. This is due to the way the `HeadLogic` is implemented and snapshot r ...
```
I am kinda anxious about the solution fix since it seems there isn't a clear plan on exactly the best way to fix this problem (rewrite was also mentioned :scream: )

ch1bo replied:
Yes, me too. Started that discussion here: https://github.com/cardano-scaling/hydra/issues/1999

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -166,21 +165,23 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
+          validBefore <- case currentPointInTime timeHandle of
+            -- XXX: We only need the current slot and this would never fail
+            Left failureReason -> throwError FailedToConstructDepositTx{failureReason}
+            Right (s, _) -> pure $ s + 200 -- XXX: configurable and unify with maxGraceTime
```
Did you pick this value since we also use it to calculate upper validity? Perhaps we need a constant to re-use somewhere or as you mention here make it configurable.

v0d1ch replied:
How could we get the current slot into scope here (if that would make get away from this error handling here.

ch1bo replied:
Hm.. yeah I should at least use maxGraceTime here. Making it configurable would actually mean making --deposit-period available here. Which I don't want to do right now.

v0d1ch replied:
Why not? I would file a separate issue to tackle this.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -277,27 +284,20 @@ instance Arbitrary OpenState where
 
 data ClosedState = ClosedState
-  { closedThreadOutput :: ClosedThreadOutput
+  { closedUTxO :: UTxO
```
I love this.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -762,33 +762,44 @@ observeInit ::
 
 -- | Observe an commit transition using a 'InitialState' and 'observeCommitTx'.
+-- NOTE: This function is a bit fragile as it assumes commit output on first
+-- output whilte the underlying observeCommitTx could deal with commit outputs
```
```suggestion
-- output while the underlying observeCommitTx could deal with commit outputs
```

ch1bo committed:
Updated spec and align some details

ch1bo committed:
Fix new/updated state change golden files and server output schemas

ch1bo committed:
Use a lower ttl in nodespec tests

ch1bo committed:
Fix generators of deposit/increment/recover txs

ch1bo committed:
Address last TODOs

ch1bo committed:
Fix flakyness in S3 tests

I have no idea why this is behaving different.

ch1bo committed:
Revert change to nothingHappensFor in behavior spec

ch1bo committed:
Remove parallel processing of ServerSpec tests

These have been increasingly flaky in CI when done in parallel. Observed
was an uncaught async exception: 'AsyncCancelled'. This was not
reproducible locally, so we decided to turn off parallel.

ch1bo committed:
Use original ttl for ReqTx and ReqDec

TxInvalid is only produced after the ttl reaches zero. Some end to end
tests (and hence users) might depend on that happening after ~500ms as
before.

ch1bo committed:
Fix imports in hydra-chain-observer

ch1bo committed:
User lower deposit period in e2e tests

ch1bo committed:
Simplify withHydraNode

ch1bo committed:
Fix upper bound and deadline on drafted deposit txs

This switches the drafted tx deadline to be 3 * deposit period.

ch1bo committed:
Reference updated spec
